llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -57,197 +57,121 @@ static cl::opt<bool>
                              cl::desc("Enable post select optimize."));
 
 namespace {
-class AIEPostSelectOptimize : public MachineFunctionPass {
-public:
-  static char ID;
-  AIEPostSelectOptimize();
-  StringRef getPassName() const override { return "AIE Post Select Optimizer"; }
-  bool runOnMachineFunction(MachineFunction &MF) override;
-  void getAnalysisUsage(AnalysisUsage &AU) const override;
 
-private:
-  const MachineRegisterInfo *MRI;
-
-  /**
-   * @struct PhysRegDefInfoStruct
-   * @brief To track the PhysReg defs and the state of their use.
-   */
-  struct PhysRegDefInfoStruct {
-    // Tracks the last visited def. Change with every new def.
-    MachineInstr *LastVisitedDef = nullptr;
-    // Tracks if the PhysReg is used in between current def and LastVisitedDef.
-    bool IsUsed = false;
-
-    PhysRegDefInfoStruct(MachineInstr *MI)
-        : LastVisitedDef(MI), IsUsed(false) {}
-  };
-
-  /// Collect all VirtReg to PhysReg copies that can be removed at the last.
-  DenseMap<Register, MachineInstr *> ErasableVirtRegToPhysRegCopies;
-
-  bool doPeepholeOpts(MachineBasicBlock &MBB);
-  bool eraseOrStoreRedundantPhysRegDefs(
-      MachineInstr &MI, DenseMap<Register, PhysRegDefInfoStruct> &PhysRegDefs);
-
-  /// Checks if the \p Op is an unallocatable phys reg.
-  bool isUnallocatablePhysReg(const MachineOperand &Op);
-
-  /// Checks if the src %virtreg of \p MI defines the same \p
-  /// PhysReg as the destination register of the \p MI.
-  bool isCopyFromPhysReg(const MachineInstr &MI, const Register PhysReg);
-
-  /// Store the first %virtreg to $physreg copy corresponding to \param MI.
-  void storeFirstIdentityCopy(MachineInstr &MI);
-
-  /// Erase \param MI if it is similar to \param LastVisitedDef.
-  bool eraseSimilarClobber(MachineInstr &MI, MachineInstr &LastVisitedDef);
+/// Information about a COPY that can be tracked by PhysRegCopyTracker.
+struct TrackableCopyOps {
+  Register VirtReg;
+  MCRegister PhysReg;
+  bool InvalidatePhysReg; // Whether PhysReg is being redefined.
```

Comment:
The comment doesn't really explain. 'is being' sounds transient. If so, describe its scope of validity

Reply:
Ok. After reading the rest: it is transient, and there's no need to store it in the map. Basically, both Phys = Virt and Virt = Phys set up an equivalence between the two which holds until it is killed. I think it would read if you would recognise both cases by their own predicates and take the appropriate actions in the main traversal.

Reply:
My point is to gather information about a COPY instruction in a single place, then it's easy to track in in the CopyTracker. I didn't want to overload the traversal and keep it at three cases:
 - Either it can be simplified
 - Or we can track the operands being copied
 - Or we can't and need to invalidate copies for any re-defined registers

Do you feel I should expose more there to make it easier to udnerstand?

Reply:
Discussed offline, I renamed `InvalidatePhysReg` to `IsPhysRegDef`